firedrake/preconditioners/fdm.py

```diff
@@ -233,71 +245,83 @@ def allocate_matrix(self, V, J, bcs, fcp, pmat_type, use_static_condensation):
-        Vsort = sorted(V, key=lambda Vsub: Vsub.dim())
         # Loop over all pairs of subspaces
-        for Vrow, Vcol in product(Vsort, Vsort):
+        for Vrow, Vcol in product(V, V):
+            if (Vrow, Vcol) in Pmats:
+                continue
+
             if symmetric and (Vcol, Vrow) in Pmats:
-                P = PETSc.Mat().createTranspose(Pmats[Vcol, Vrow])
-            else:
-                on_diag = Vrow == Vcol
-                triu = on_diag and symmetric
-                ptype = pmat_type if on_diag else PETSc.Mat.Type.AIJ
-                sizes = tuple(Vsub.dof_dset.layout_vec.getSizes() for Vsub in (Vrow, Vcol))
-
-                preallocator = PETSc.Mat().create(comm=self.comm)
-                preallocator.setType(PETSc.Mat.Type.PREALLOCATOR)
-                preallocator.setSizes(sizes)
-                preallocator.setOption(PETSc.Mat.Option.IGNORE_ZERO_ENTRIES, False)
-                preallocator.setUp()
-                self.set_values(preallocator, Vrow, Vcol, addv, triu=triu)
-
-                preallocator.assemble()
-                d_nnz, o_nnz = get_preallocation(preallocator, sizes[0][0])
-                preallocator.destroy()
-                if on_diag:
-                    numpy.maximum(d_nnz, 1, out=d_nnz)
-
-                P = PETSc.Mat().create(comm=self.comm)
-                P.setType(ptype)
-                P.setSizes(sizes)
-                P.setPreallocationNNZ((d_nnz, o_nnz))
-                P.setOption(PETSc.Mat.Option.NEW_NONZERO_ALLOCATION_ERR, True)
-                if on_diag:
-                    P.setOption(PETSc.Mat.Option.STRUCTURALLY_SYMMETRIC, True)
-                if ptype.endswith("sbaij"):
-                    P.setOption(PETSc.Mat.Option.IGNORE_LOWER_TRIANGULAR, True)
-                P.setUp()
-                # append callables to zero entries, insert element matrices, and apply BCs
-                assembly_callables.append(P.zeroEntries)
-                assembly_callables.append(partial(self.set_values, P, Vrow, Vcol, addv))
-                if on_diag:
-                    own = Vrow.dof_dset.layout_vec.getLocalSize()
-                    bdofs = numpy.flatnonzero(self.lgmaps[Vrow].indices[:own] < 0).astype(PETSc.IntType)[:, None]
-                    Vrow.dof_dset.lgmap.apply(bdofs, result=bdofs)
-                    if len(bdofs) > 0:
-                        vals = numpy.ones(bdofs.shape, dtype=PETSc.RealType)
-                        assembly_callables.append(partial(P.setValuesRCV, bdofs, bdofs, vals, addv))
+                Pmats[Vrow, Vcol] = PETSc.Mat().createTranspose(Pmats[Vcol, Vrow])
+                continue
+
+            # Preallocate and assemble the FDM auxiliary sparse operator
+            on_diag = Vrow == Vcol
+            sizes = tuple(Vsub.dof_dset.layout_vec.getSizes() for Vsub in (Vrow, Vcol))
+            ptype = pmat_type if on_diag else PETSc.Mat.Type.AIJ
+
+            preallocator = PETSc.Mat().create(comm=self.comm)
```

Comment:
Can this not be achieved by using PyOP2 `Sparsity`? Why are you hand-rolling this matrix instantiation code?

Reply:
Does that allow sparse element matrices or just dense blocks?

Reply:
No. I now see what's going on.